linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/Emitter.scala

```diff
@@ -715,7 +715,7 @@ final class Emitter[E >: Null <: js.Tree](
             )(moduleContext, classCache, linkedClass.pos).map(postTransform(_, 0))))
       }
 
-      if (linkedClass.hasInstances && kind.isClass && linkedClass.hasRuntimeTypeInfo) {
+      if (linkedClass.hasDirectInstances && kind.isClass) {
```

Comment:
Just to check my understanding: `hasDirectInstances` implies `hasRuntimeTypeInfo` because
- we need it for arrays and we do not DCE arrays
- we need it for jl.Object `toString` which is always reachable

Maybe this is worth a comment? (I'm not sure here is the right place though). :-/

Alternatively, leave the guard in place just to make it less confusing.

Reply:
The Analyzer has a direct implication: `isInstantiated` directly implies `dataAccessed`.

I don't think arrays come into play here. The `.prototype.$classData` of arrays take a completely different path, in `CoreJSLib`. This code path only applies to regular classes.

> * we need it for jl.Object `toString` which is always reachable

That's one reason, but there is a more fundamental reason. Interface instance tests of the form `x.isInstanceOf[Intf]` rely on `x.$classData.ancestors`, and that goes through `$c_ClassOfX.prototype.$classData`. So even if `toString()` were no longer reachable, we would still need the run-time type info of every directly instantiated class.